pkg/collector/corechecks/servicediscovery/usm/ruby.go

```diff
@@ -0,0 +1,96 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package usm
+
+import (
+	"io"
+	"io/fs"
+	"regexp"
+	"strings"
+
+	"github.com/shirou/gopsutil/v3/process"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+var (
+	moduleRegexp  = regexp.MustCompile(`module\s+([A-Z][a-zA-Z0-9_]*)`)
+	matchFirstCap = regexp.MustCompile("(.)([A-Z][a-z]+)")
+	matchAllCap   = regexp.MustCompile("([a-z0-9])([A-Z])")
+)
+
+type railsDetector struct {
+	ctx DetectionContext
+}
+
+func newRailsDetector(ctx DetectionContext) detector {
+	return &railsDetector{ctx}
+}
+
+// detect checks if the service is a Rails application by looking for a
+// `config/application.rb` file generated by `rails new` when a new rails
+// project is created. This file should contain a `module` declaration with the
+// application name.
+func (r railsDetector) detect(_ []string) (ServiceMetadata, bool) {
+	proc := r.ctx.contextMap[ServiceProc].(*process.Process)
```

Comment:
break it down to multiple-safe-calls
never rely on a map to contain a value, and never blindly convert the returned type to a different type. That can lead to panic in production